src/OrchardCore.Modules/OrchardCore.Media/SecureMediaPermissions.cs

```diff
@@ -0,0 +1,155 @@
+    {
+        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
+        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
+        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
+
+        private static readonly Permission s_viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in {0}", new[] { ViewMedia });
+
+        private static Dictionary<ValueTuple<string, string>, Permission> s_permissionsByFolder = new();
+        private static readonly char[] s_trimSecurePathChars = ['/', '\\', ' '];
+        public static readonly ReadOnlyDictionary<string, Permission> PermissionTemplates = new(new Dictionary<string, Permission>()
+        {
+            { ViewMedia.Name, s_viewMediaTemplate },
+        });
+
+        private readonly MediaOptions _mediaOptions;
+        private readonly AttachedMediaFieldFileService _attachedMediaFieldFileService;
+        private readonly IMediaFileStore _fileStore;
+
+        public SecureMediaPermissions(
+            IOptions<MediaOptions> options,
+            IMediaFileStore fileStore,
+            AttachedMediaFieldFileService attachedMediaFieldFileService)
+        {
+            _mediaOptions = options.Value;
+            _fileStore = fileStore;
+            _attachedMediaFieldFileService = attachedMediaFieldFileService;
+        }
+
+        public async Task<IEnumerable<Permission>> GetPermissionsAsync()
+        {
+            // The ViewRootMedia permission must be implied by any sub folder permission.
+            var viewRootImpliedBy = new List<Permission>(ViewRootMedia.ImpliedBy);
+
+            var result = new List<Permission>()
+            {
+                ViewMedia,
+                new (ViewRootMedia.Name, ViewRootMedia.Description, viewRootImpliedBy),
+                ViewOthersMedia,
+                ViewOwnMedia
+            };
+
+            await foreach (var entry in _fileStore.GetDirectoryContentAsync())
+            {
+                if (!entry.IsDirectory)
+                    continue;
+
+                if (entry.Name == _mediaOptions.AssetsUsersFolder ||
+                    entry.Name == _attachedMediaFieldFileService.MediaFieldsFolder)
+                    continue;
+
+                var folderPath = entry.Path;
+
+                foreach (var template in PermissionTemplates)
+                {
+                    var dynamicPermission = CreateDynamicPermission(template.Value, folderPath);
```

Comment:
Do you have an idea for supporting folder renames?

Reply:
Currently, renaming a folder creates a new permission and will require you to manually assign it to a role. I don't think we can detect this case with the existing file system abstraction.

Reply:
Perhaps display a notification when renaming secured folders?

Reply:
At the moment, renaming folders is not possible in the media library anyway. Or did I miss something?

Reply:
No, indeed, just moving files.